ksql-engine/src/main/java/io/confluent/ksql/connect/ConnectPollingService.java

```diff
@@ -53,41 +58,85 @@
  * registry.</p>
  */
 @ThreadSafe
-final class ConnectPollingService extends AbstractScheduledService {
+final class ConnectPollingService extends AbstractExecutionThreadService  {
 
   private static final Logger LOG = LoggerFactory.getLogger(ConnectPollingService.class);
-  private static final int INTERVAL_S = 30;
+  private static final int MAX_INTERVAL_S = 30;
+  private static final int START_INTERVAL_S = 1;
+  private static final Connector STOP_SENTINEL =
+      new Connector("_stop_", ignored -> false, Function.identity(), DataSourceType.KSTREAM, "");
 
   private final KsqlExecutionContext executionContext;
   private final Consumer<CreateSource> sourceCallback;
+  private final int maxPollingIntervalSecs;
+  private final AtomicInteger pollingIntervalSecs;
 
+  // we use a blocking queue as a thread safe buffer between this class
+  // and others that may be calling #addConnector(Connector) - this also
+  // allows us to notify when a connector was added.
+  private BlockingQueue<Connector> connectorQueue;
   private Set<Connector> connectors;
 
   ConnectPollingService(
       final KsqlExecutionContext executionContext,
       final Consumer<CreateSource> sourceCallback
+  ) {
+    this(executionContext, sourceCallback, MAX_INTERVAL_S);
+  }
+
+  @VisibleForTesting
+  ConnectPollingService(
+      final KsqlExecutionContext executionContext,
+      final Consumer<CreateSource> sourceCallback,
+      final int maxPollingIntervalSecs
   ) {
     this.executionContext = Objects.requireNonNull(executionContext, "executionContext");
     this.sourceCallback = Objects.requireNonNull(sourceCallback, "sourceCallback");
-    this.connectors = ConcurrentHashMap.newKeySet();
+    this.connectors = new HashSet<>();
+    this.connectorQueue = new LinkedBlockingDeque<>();
+    this.maxPollingIntervalSecs = maxPollingIntervalSecs;
+    this.pollingIntervalSecs = new AtomicInteger(maxPollingIntervalSecs);
   }
 
   /**
    * Add this connector to the set of connectors that are polled by this
    * {@code ConnectPollingService}. Next time an iteration is scheduled in
-   * {@value #INTERVAL_S} seconds, the connector will be included in the topic
+   * {@value #MAX_INTERVAL_S} seconds, the connector will be included in the topic
    * scan.
    *
    * @param connector a connector to register
    */
   void addConnector(final Connector connector) {
-    connectors.add(connector);
+    connectorQueue.add(connector);
+    pollingIntervalSecs.set(START_INTERVAL_S);
   }
 
   @Override
-  protected void runOneIteration() {
+  protected void run() throws Exception {
+    while (isRunning()) {
+      final Connector connector = connectorQueue.poll(
+          pollingIntervalSecs.getAndUpdate(old -> Math.max(MAX_INTERVAL_S, 2 * old)),
```

Comment:
This looks like it's supposed to be `Math.min(...)` instead?

Reply:
oops.